src/array.c

```diff
@@ -629,156 +642,408 @@ static size_t limit_overallocation(jl_array_t *a, size_t alen, size_t newlen, si
+                                        size_t inc, size_t n)
 {
-    if (a->flags.isshared && a->flags.how!=3) jl_error("cannot resize array with shared data");
     // optimized for the case of only growing and shrinking at the end
-    size_t alen = jl_array_nrows(a);
-    if ((alen + inc) > a->maxsize - a->offset) {
-        size_t newlen = a->maxsize==0 ? (inc<4?4:inc) : a->maxsize*2;
-        while ((alen + inc) > newlen - a->offset)
+    if (__unlikely(a->flags.isshared)) {
+        if (a->flags.how != 3)
+            jl_error("cannot resize array with shared data");
+        if (inc == 0) {
+            // If inc > 0, it will always trigger the slow path and unshare the
+            // buffer
+            array_try_unshare(a);
+            return;
+        }
+    }
+    size_t elsz = a->elsize;
+    char *data = (char*)a->data;
+    int has_gap = n > idx;
+    if (__unlikely((n + inc) > a->maxsize - a->offset)) {
+        size_t nb1 = idx * elsz;
+        size_t nbinc = inc * elsz;
+        size_t newlen = a->maxsize == 0 ? (inc < 4 ? 4 : inc) : a->maxsize * 2;
+        while ((n + inc) > newlen - a->offset)
             newlen *= 2;
-
-        newlen = limit_overallocation(a, alen, newlen, inc);
-        array_resize_buffer(a, newlen, alen, a->offset);
+        newlen = limit_overallocation(a, n, newlen, inc);
+        int newbuf = array_resize_buffer(a, newlen);
+        char *newdata = (char*)a->data + a->offset * elsz;
+        if (newbuf) {
+            memcpy(newdata, data, nb1);
+            if (has_gap) {
+                memcpy(newdata + nb1 + nbinc, data + nb1, n * elsz - nb1);
+            }
+        }
+        else if (has_gap) {
+            memmove(newdata + nb1 + nbinc, newdata + nb1, n * elsz - nb1);
+        }
+        a->data = data = newdata;
     }
+    else if (has_gap) {
+        size_t nb1 = idx * elsz;
+        memmove(data + nb1 + inc * elsz, data + nb1, n * elsz - nb1);
+    }
+    size_t newnrows = n + inc;
 #ifdef STORE_ARRAY_LEN
-    a->length += inc;
+    a->length = newnrows;
 #endif
-    a->nrows += inc;
+    a->nrows = newnrows;
+    if (a->flags.ptrarray) {
+        memset(data + idx * elsz, 0, inc * elsz);
+    }
 }
 
-JL_DLLEXPORT void jl_array_del_end(jl_array_t *a, size_t dec)
+JL_DLLEXPORT void jl_array_grow_at(jl_array_t *a, ssize_t idx, size_t inc)
 {
-    if (dec == 0) return;
-    if (dec > a->nrows)
-        jl_bounds_error_int((jl_value_t*)a, a->nrows - dec);
-    if (a->flags.isshared) array_try_unshare(a);
-    if (a->elsize > 0) {
-        char *ptail = (char*)a->data + (a->nrows-dec)*a->elsize;
-        assert(ptail < (char*)a->data + (a->length*a->elsize));
-        if (a->flags.ptrarray)
-            memset(ptail, 0, dec*a->elsize);
-        else
-            ptail[0] = 0;
+    // No need to explicitly unshare.
+    // Shared arrays are guaranteed to trigger the slow path for growing.
+    size_t n = jl_array_nrows(a);
+    if (idx < 0 || idx > n)
```

Comment:
No, it is allowed to grow at the end of the array (`jl_array_grow_end` below also calls `jl_array_grow_at_end` with the `inc == n`). In general, there are `n + 1` positions one can insert elements in an `n` elements array, which are represented as `0` to `n` here.